task_download.py

```diff
@@ -0,0 +1,147 @@
+import argparse
+import requests
+import json
+import os
+import logging
+from concurrent.futures import ThreadPoolExecutor
+from datetime import datetime
+
+# constants
+BATCH_URL = 'https://api.scale.com/v1/batches'
+TASK_URL = 'https://api.scale.com/v1/tasks'
+MAX_WORKERS = 200
+
+# globals
+auth = None
+project = None
+batch_total = 0
+batch_count = 0
+task_count = 0
+error_batches = []
+
+# create logs and batches directories
+if not os.path.isdir('task_download'):
+    os.mkdir('task_download')
+if not os.path.isdir('task_download/logs'):
+    os.mkdir('task_download/logs')
+if not os.path.isdir('task_download/batches'):
+    os.mkdir('task_download/batches')
+
+# logger
+logging.basicConfig(format='%(levelname)s %(message)s', level=logging.INFO)
+info = logging.info
+error = logging.error
+logger = logging.getLogger()
+logger.addHandler(logging.FileHandler(
+    f'task_download/logs/{datetime.now().strftime("%Y%m%d_%H%M%S")}.log'))
+
+
+def main():
+    global batch_total, auth, project
+
+    args = get_args()
+    auth = (args.api_key, '')
+    resume = args.resume
+    project = args.project
+
+    batches = get_batches()
+    batch_total = len(batches)
+
+    info(f'Downloading {batch_total} batches...')
+
+    futures = []
+    threadpool = ThreadPoolExecutor(max_workers=MAX_WORKERS)
+    for batch in batches:
+        if resume and os.path.isfile(f'task_download/batches/{batch}.json'):
+            batch_total -= 1
+            info(f'Skipping batch {batch} since --resume flag was used')
+        else:
+            future = threadpool.submit(get_tasks, batch)
+            futures.append(future)
+
+    for future in futures:
+        future.result()
+
+    if(len(error_batches)):
+        error(f'Error batches: {error_batches}')
+        error(
+            f'There were errors downloading {len(error_batches)} batches, to retry run the script with the --resume flag')
+
+
+def get_batches():
+    batches = []
+    has_next_page = True
+    next_page = 1
+    info('Getting batch pages...')
+    try:
+        while has_next_page:
+            params = {'status': 'completed', 'offset': (next_page - 1) * 100}
```

Comment:
I might make the params either an arg, or declared explicitly at the top of the file as something that can be changed. Then set `offset` as just a key within the dict here.

Reply:
Use-case: Customers may want to retrieve all tasks regardless of status, or only tasks created between dates, that have an accepted audit, etc.